autoload/ale/engine.vim

```diff
@@ -184,36 +184,42 @@ endfunction
 function! ale#engine#SetResults(buffer, loclist) abort
     let l:linting_is_done = !ale#engine#IsCheckingBuffer(a:buffer)
 
-    " Set signs first. This could potentially fix some line numbers.
-    " The List could be sorted again here by SetSigns.
-    if g:ale_set_signs
-        call ale#sign#SetSigns(a:buffer, a:loclist)
-    endif
+    if g:ale_send_to_neovim_diagnostics
+        call ale#engine#SendResultsToNeovimDiagnostics(a:buffer, a:loclist)
+    else
+        " Set signs first. This could potentially fix some line numbers.
+        " The List could be sorted again here by SetSigns.
+        if g:ale_set_signs
+            call ale#sign#SetSigns(a:buffer, a:loclist)
+        endif
+
+        if g:ale_set_quickfix || g:ale_set_loclist
+            call ale#list#SetLists(a:buffer, a:loclist)
+        endif
 
-    if g:ale_set_quickfix || g:ale_set_loclist
-        call ale#list#SetLists(a:buffer, a:loclist)
+        if g:ale_set_highlights
+            call ale#highlight#SetHighlights(a:buffer, a:loclist)
+        endif
     endif
 
     if exists('*ale#statusline#Update')
         " Don't load/run if not already loaded.
         call ale#statusline#Update(a:buffer, a:loclist)
     endif
 
-    if g:ale_set_highlights
-        call ale#highlight#SetHighlights(a:buffer, a:loclist)
-    endif
-
     if l:linting_is_done
-        if g:ale_echo_cursor
-            " Try and echo the warning now.
-            " This will only do something meaningful if we're in normal mode.
-            call ale#cursor#EchoCursorWarning()
-        endif
+        if !g:ale_send_to_neovim_diagnostics
+            if g:ale_echo_cursor
+                " Try and echo the warning now.
+                " This will only do something meaningful if we're in normal mode.
+                call ale#cursor#EchoCursorWarning()
+            endif
 
-        if g:ale_virtualtext_cursor
-            " Try and show the warning now.
-            " This will only do something meaningful if we're in normal mode.
-            call ale#virtualtext#ShowCursorWarning()
+            if g:ale_virtualtext_cursor
+                " Try and show the warning now.
+                " This will only do something meaningful if we're in normal mode.
+                call ale#virtualtext#ShowCursorWarning()
```

Comment:
I think we should keep the `g:ale_echo_cursor` warnings even if the Neovim API is enabled. That way you're sure to see problems by default on lines if something goes wrong with the rest of Neovim. All that does is show a message when your cursor gets to a line, and that might not be something that Neovim will do for you. That's my suggestion.

If people don't like it, they can always add in `let g:ale_echo_cursor = 0` in addition to the new setting. We need to document whatever behaviour we go with. I think `g:ale_virtualtext_cursor` ought to be ignored when sending diagnostics to NeoVim.